proposals/NNNN-opt-in-reflection-metadata.md

```diff
@@ -0,0 +1,181 @@
+
+
+## Detailed design
+
+Since Reflection metadata might be used by the debugger, we propose to always keep that metadata if full emission of debugging information is enabled (`-gdwarf-types` or `-g` flags).
+
+**Changes in flags**
+To handle behavior change between Swift pre-6 and 6, we can introduce a new upcoming feature, which will allow to enable Opt-In mode explicitly for pre-6 Swift with `-enable-upcoming-feature OptInReflection` and will set this mode by default in Swift 6.
+
+A new flag `-enable-full-reflection-metadata` will also have to be introduced to allow developers to enable reflection in full if they desire in Swift 6 and later.
+
+For Swift 6, flags `-disable-reflection-metadata` and `-emit-reflection-for-debugger` will be a no-op, to ensure the reflection metadata is always available when needed.
+
+1.  Reflection Disabled (`-disable-reflection-metadata` and `-reflection-metadata-for-debugger-only`)
+- For Swift pre-6 emit Reflection metadata only if full debugging information is enabled.
+- If there is a type in a module conforming to `Reflectable`, the compiler will emit an error.
+- A no-op in Swift 6 and later (Opt-in mode is enabled by default).
+
+2.  Opt-In Reflection (`-enable-upcoming-feature OptInReflection`)
+- If debugging is disabled, emit only for types that conform to `Reflectable`.
+- Will emit reflection in full for all types if debugging is enabled.
+- For Swift pre-6 will require an explicit flag, for Swift 6 will be enabled by default.
+
+3.  Fully enabled (`-enable-full-reflection-metadata`)
+- Always emit reflection metadata for all types regardless of debugging information level.
+- Conformance to Reflectable will be synthesized for all types to allow usage of reflection-consuming APIs.
+- Current default level for Swift pre-6.
+
+Introducing a new flag to control the feature will allow us to safely roll it out and avoid breakages of the existing code. For those modules that get compiled with fully enabled metadata, nothing will change (all symbols will stay present). For modules that have the metadata disabled, but are consumers of reflectable API, the compiler will emit the error enforcing the guarantee.
+
+### Casts implementation
+Casting might be a good way to improve the feature's ergonomics because currently there is no way to check if reflection is available at runtime. (`Mirror.children.count` doesn't really help because it doesn't distinguish between the absence of reflection metadata and the absence of fields on a type)
+
+To implement this feature, we propose to introduce a new runtime function `swift_reflectableCast`, and emit a call to it instead of `swift_dynamicCast`during IRGen if Reflectable is a target type.
```

Comment:
Should we handle this in `swift_dynamicCast` as well?

Reply:
I considered handling all casting logic in `swift_dynamicCast `, but since `Reflectable` is a marker protocol it has no runtime representation and hence `swift_dynamicCast` can't know about it.

Reply:
So to handle that case `swift_dynamicCast` we would need to introduce a new option and pass it to the function. However, it would make the logic of `swift_dynamicCast` more complicated and we would still need to calculate that flag at IRGen.

Reply:
Good point.  I had missed that `Reflectable` had no runtime representation.